runtime/compiler/aarch64/codegen/J9AheadOfTimeCompile.cpp

```diff
@@ -24,14 +24,71 @@
 #include "codegen/CodeGenerator.hpp"
 
 J9::ARM64::AheadOfTimeCompile::AheadOfTimeCompile(TR::CodeGenerator *cg) :
-         J9::AheadOfTimeCompile(_relocationTargetTypeToHeaderSizeMap, cg->comp()),
-         _cg(cg)
+      J9::AheadOfTimeCompile(_relocationTargetTypeToHeaderSizeMap, cg->comp()),
+      _cg(cg),
+      _relocationList(getTypedAllocator<TR::ARM64Relocation*>(cg->comp()->allocator()))
    {
    }
 
 void J9::ARM64::AheadOfTimeCompile::processRelocations()
    {
-   TR_UNIMPLEMENTED();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(_cg->fe());
+   TR::IteratedExternalRelocation *r;
+
+   for (auto iterator = getRelocationList().begin(); iterator != getRelocationList().end(); ++iterator)
+      (*iterator)->mapRelocation(_cg);
+
+   for (auto aotIterator = _cg->getExternalRelocationList().begin(); aotIterator != _cg->getExternalRelocationList().end(); ++aotIterator)
+      {
+      (*aotIterator)->addExternalRelocation(_cg);
+      }
+
+   for (r = getAOTRelocationTargets().getFirst(); r != NULL; r = r->getNext())
+      {
+      addToSizeOfAOTRelocations(r->getSizeOfRelocationData());
+      }
+
+   // now allocate the memory  size of all iterated relocations + the header (total length field)
+
+   // Note that when using the SymbolValidationManager, the well-known classes
+   // must be checked even if no explicit records were generated, since they
+   // might be responsible for the lack of records.
+   bool useSVM = self()->comp()->getOption(TR_UseSymbolValidationManager);
+
+   if (self()->getSizeOfAOTRelocations() != 0 || useSVM)
+      {
+      // It would be more straightforward to put the well-known classes offset
+      // in the AOT method header, but that would use space for AOT bodies that
+      // don't use the SVM.
+      int wellKnownClassesOffsetSize = useSVM ? SIZEPOINTER : 0;
+      uintptrj_t reloBufferSize =
+         self()->getSizeOfAOTRelocations() + SIZEPOINTER + wellKnownClassesOffsetSize;
+      uint8_t *relocationDataCursor =
+         self()->setRelocationData(fej9->allocateRelocationData(self()->comp(), reloBufferSize));
+
+      // set up the size for the region
+      *(uint32_t *)relocationDataCursor = reloBufferSize;
+      relocationDataCursor += SIZEPOINTER;
+
+      if (useSVM)
+         {
+         TR::SymbolValidationManager *svm =
+            self()->comp()->getSymbolValidationManager();
+         void *offsets = const_cast<void*>(svm->wellKnownClassChainOffsets());
+         *(uintptr_t *)relocationDataCursor = reinterpret_cast<uintptr_t>(
+            fej9->sharedCache()->offsetInSharedCacheFromPointer(offsets));
+         relocationDataCursor += SIZEPOINTER;
+         }
+
+      // set up pointers for each iterated relocation and initialize header
+      TR::IteratedExternalRelocation *s;
+      for (s = getAOTRelocationTargets().getFirst(); s != NULL; s = s->getNext())
+         {
+         s->setRelocationData(relocationDataCursor);
+         s->initializeRelocation(_cg);
```

Comment:
Note, this will result in a call to `J9::ARM64::AheadOfTimeCompile::initializeAOTRelocationHeader` which is unimplemented at the moment.

Reply:
`initializeAOTRelocationHeader()` is implemented in #7222.